Clean up dangling images

B1F030 opened the pull request:
**What type of PR is this?**
/kind cleanup
<!--
Add one of the following kinds:

/kind api-change
/kind bug
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind feature
/kind flake

-->

**What this PR does / why we need it**:
Clean up dangling images after each run of `hack/local-up-karmada.sh`.

**Which issue(s) this PR fixes**:
Fixes #4914

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required.
-->
```release-note
NONE
```

B1F030 commented:
The result of this PR change are reflected [here](https://github.com/karmada-io/karmada/actions/runs/8996981817/job/24714449839?pr=4915#step:5:695).
```
Total reclaimed space: 0B
```
If we try this on normal environment, this change could clean up a lot of storage space.

B1F030 commented:
> I tried on my side, it shows that reclaimed 0B space, with a bunch of images deleted, and I don't know why :)

Sorry for my mistake, cleaning up dangling images may not reclaim storage space, if you have the same image tagged.
For example:
```
karmada/karmada-aggregated-apiserver          latest    93d7db7c82c7   14 minutes ago      98.2MB
<none>                                        <none>    d7207b9f233d   About an hour ago   98.2MB
```
These two images are actually the same, using the same filesystem, and having same layers, just have different id because they are built from different run.
In this case, the `Total reclaimed space` is correct to be `0`.

zhzhuang-zju commented:
> In this case, the Total reclaimed space is correct to be 0.

I verified locally that deleting the dangling image generated by command `docker build .`, the `Total reclaimed space` shown is also 0. Many related issues can be found, such as https://github.com/docker/cli/issues/1751, but it seems that there is no answer yet

B1F030 commented:
```
appledeMacBook-Pro:~ apple$ docker images
REPOSITORY                             TAG       IMAGE ID       CREATED          SIZE
karmada/karmada-aggregated-apiserver   latest    688235b8f0fb   4 seconds ago    98.7MB
<none>                                 <none>    4d63cb6d9596   25 seconds ago   98.7MB
```
I tried remove `karmada/karmada-aggregated-apiserver` first, and then `docker image prune -f`, still got `Total reclaimed space: 0B`. Weird, but seems to be a bug of docker.

XiShanYongYe-Chang commented on hack/local-up-karmada.sh:
```diff
@@ -132,6 +132,8 @@ util::create_cluster "${PULL_MODE_CLUSTER_NAME}" "${PULL_MODE_CLUSTER_TMP_CONFIG
 export REGISTRY="docker.io/karmada"
 make images GOOS="linux" --directory="${REPO_ROOT}"
+#clean up dangling images
+docker image prune -f
```
What is the usage scenario of the dancing image?

B1F030 replied:
What about add `LABEL image-source=karmada` when docker build, so that we can only remove dangling images created by karmada by `docker image prune -f --filter="label=image-source=karmada"`

B1F030 replied:
> Excellent ideal! And furtherly, label it with a uuid, guarantee only clean up images made in this executaion.

I think there's no need to use uuid, just `LABEL maintainer="karmada.io"` should work fine. WDYT?

B1F030 commented:
```
appledeMacBook-Pro:~ apple$ docker images
REPOSITORY                             TAG       IMAGE ID       CREATED              SIZE
karmada/karmada-aggregated-apiserver   latest    ce11352b8dd2   24 seconds ago       98.7MB
<none>                                 <none>    f0da7dd4ea10   About a minute ago   98.7MB
appledeMacBook-Pro:~ apple$ 
appledeMacBook-Pro:~ apple$ docker inspect f0da7dd4ea10
{...
            "Labels": {
                "maintainer": "karmada.io"
            }
...}
appledeMacBook-Pro:~ apple$ 
appledeMacBook-Pro:~ apple$ docker inspect ce11352b8dd2
{...
            "Labels": {
                "maintainer": "karmada.io"
            }
...}
appledeMacBook-Pro:~ apple$ 
appledeMacBook-Pro:~ apple$ docker image prune --force --filter "label=maintainer=karmada.io"
Deleted Images:
deleted: sha256:f0da7dd4ea10629d0205c320e17e91df9183238ff51f1d347366dd78cc47f55a

Total reclaimed space: 0B
appledeMacBook-Pro:~ apple$ 
appledeMacBook-Pro:~ apple$ docker images
REPOSITORY                             TAG       IMAGE ID       CREATED          SIZE
karmada/karmada-aggregated-apiserver   latest    ce11352b8dd2   34 seconds ago   98.7MB
```
Works in my local environment.

zhzhuang-zju commented on cluster/images/Dockerfile:
```diff
@@ -16,6 +16,8 @@ FROM alpine:3.19.1
 
 ARG BINARY
 
+LABEL maintainer="karmada.io"
```
@B1F030 Can `LABEL` refer to `REGISTRY` and `VERSION` as the input environment variables of `docker.sh` , so that different scenarios can be configured as needed, and the scope of the label can also be reduced.

```shell
function build_local_image() {
  local -r output_type=$1
  local -r target=$2
  local -r platform=$3

  local -r image_name="${REGISTRY}/${target}:${VERSION}"

  echo "Building image for ${platform}: ${image_name}"
  set -x
  docker build --build-arg BINARY="${target}" \
          ${DOCKER_BUILD_ARGS} \
          --tag "${image_name}" \
          --file "${REPO_ROOT}/cluster/images/Dockerfile" \
          --label "${LABEL}"\
          "${REPO_ROOT}/_output/bin/${platform}"
  set +x

  if [[ "$output_type" == "registry" ]]; then
    docker push "${image_name}"
    signImage ${image_name}
  fi
}
```

B1F030 replied:
What about now?
And in my humble opinion, we should clean dangling images each run, so no need to match the label exactly in docker prune.

B1F030 replied:
Updated to `--label "version.karmada.io=${REGISTRY}:${VERSION}"`.

zhzhuang-zju commented on hack/local-up-karmada.sh:
```diff
@@ -131,7 +131,10 @@ util::create_cluster "${PULL_MODE_CLUSTER_NAME}" "${PULL_MODE_CLUSTER_TMP_CONFIG
 #step2. make images and get karmadactl
 export VERSION="latest"
 export REGISTRY="docker.io/karmada"
+export LABEL="repository=karmada.io"
```
```suggestion
export KARMADA_IMAGE_LABEL="repository=karmada.io"
```
Adding karmada to the variable's name may avoid conflicts with existing variables on the environment.

B1F030 replied:
SGTM, done.

zhzhuang-zju commented on hack/docker.sh:
```diff
@@ -47,6 +47,7 @@ source "${REPO_ROOT}/hack/util.sh"
 
 REGISTRY=${REGISTRY:-"docker.io/karmada"}
 VERSION=${VERSION:="unknown"}
+LABEL=${LABEL:-}
```
```suggestion
LABEL=${KARMADA_IMAGE_LABEL:-}
```

zhzhuang-zju commented on hack/docker.sh:
```diff
@@ -78,6 +79,7 @@ function build_local_image() {
           ${DOCKER_BUILD_ARGS} \
           --tag "${image_name}" \
           --file "${REPO_ROOT}/cluster/images/Dockerfile" \
+          --label "${LABEL}" \
```
Need to deal with the scenario if the label is empty

B1F030 commented:
/cc @zhzhuang-zju

zhzhuang-zju commented:
> /cc @zhzhuang-zju

Good job! LGTM

cc @ikaven1024

zhzhuang-zju commented on hack/local-up-karmada.sh:
```diff
@@ -131,7 +131,10 @@ util::create_cluster "${PULL_MODE_CLUSTER_NAME}" "${PULL_MODE_CLUSTER_TMP_CONFIG
 #step2. make images and get karmadactl
 export VERSION="latest"
 export REGISTRY="docker.io/karmada"
+export KARMADA_IMAGE_LABEL="karmada.io"
```
> you can set it in DOCKER_BUILD_ARGS variable, and no need modify docker.sh.

Setting labels in `DOCKER_BUILD_ARGS` do work functionally. But the variable `DOCKER_BUILD_ARGS` is used to carry `--build-arg`, which is the same level as `--label`. Directly setting labels in `DOCKER_BUILD_ARGS` may not be appropriate.
```bash
(base) ➜  ksense git:(setup) ✗ docker build --help
...
Options:
      --build-arg stringArray         Set build-time variables
      --label stringArray             Set metadata for an image
```

zhzhuang-zju replied:
> Should we give a more exact label name than `karmada.io`. Personally it seems confusing.

Adding comments might be a little better. Do you have any ideas about label value?

zhzhuang-zju replied:
> `DOCKER_BUILD_ARGS` should be carried any arguments, not only `--build-arg`.

If so, setting labels in DOCKER_BUILD_ARGS would be fine.

>  would you test it. If some error occur, you can show it to me.

@B1F030 has tested it and there is nothing functionally wrong with it

B1F030 replied:
Seems cannot use "May be pruned in local-up-karmada.sh" as an argument in sctipt, may cause some error, so I insert some "_" into the value.

B1F030 commented:
/retest

zhzhuang-zju commented:
/ok-to-test

B1F030 commented:
/retest

B1F030 commented:
@RainbowMango PTAL, thanks.

RainbowMango commented:
/assign

RainbowMango reviewed (approved):
/lgtm
/approve

Thanks for the excellent efforts on this @ikaven1024 @zhzhuang-zju @B1F030

RainbowMango commented on hack/local-up-karmada.sh:
```diff
@@ -131,7 +131,11 @@ util::create_cluster "${PULL_MODE_CLUSTER_NAME}" "${PULL_MODE_CLUSTER_TMP_CONFIG
 export VERSION="latest"
 export REGISTRY="docker.io/karmada"
+export KARMADA_IMAGE_LABEL_VALUE="May_be_pruned_in_local-up-karmada.sh"
+export DOCKER_BUILD_ARGS="--label=image.karmada.io=${KARMADA_IMAGE_LABEL_VALUE}"
```
Got, no big deal.
Would you like to send a PR for this?

B1F030 replied:
> This script should be:
> 
> ```shell
> export DOCKER_BUILD_ARGS="${DOCKER_BUILD_ARGS} --label=image.karmada.io=${KARMADA_IMAGE_LABEL_VALUE}"
> ```

It seems we can not call the `${DOCKER_BUILD_ARGS}` here, since it has never been defined before.

B1F030 replied:
Fixed in #4939, PTAL.
